app/controllers/api/vcs.php

```diff
@@ -0,0 +1,1019 @@
+use Utopia\Detector\Adapter\Java;
+use Utopia\Detector\Adapter\JavaScript;
+use Utopia\Detector\Adapter\PHP;
+use Utopia\Detector\Adapter\Python;
+use Utopia\Detector\Adapter\Ruby;
+use Utopia\Detector\Adapter\Swift;
+use Utopia\Detector\Detector;
+use Utopia\Validator\Boolean;
+
+use function Swoole\Coroutine\batch;
+
+App::get('/v1/vcs/github/installations')
+    ->desc('Install GitHub App')
+    ->groups(['api', 'vcs'])
+    ->label('scope', 'public')
+    ->label('origin', '*')
+    ->label('sdk.auth', [])
+    ->label('sdk.namespace', 'vcs')
+    ->label('sdk.method', 'createGitHubInstallation')
+    ->label('sdk.description', '')
+    ->label('sdk.response.code', Response::STATUS_CODE_MOVED_PERMANENTLY)
+    ->label('sdk.response.type', Response::CONTENT_TYPE_HTML)
+    ->label('sdk.methodType', 'webAuth')
+    ->param('redirect', '', fn ($clients) => new Host($clients), 'URL to redirect back to your Git authorization. Only console hostnames are allowed.', true, ['clients'])
+    ->param('projectId', '', new UID(), 'Project ID')
+    ->inject('response')
+    ->inject('user')
+    ->inject('dbForConsole')
+    ->action(function (string $redirect, string $projectId, Response $response, Document $user, Database $dbForConsole) {
+        $state = \json_encode([
+            'projectId' => $projectId,
+            'redirect' => $redirect
+        ]);
+
+        // replace github url state with vcsState in user prefs attribute
+        $prefs = $user->getAttribute('prefs', []);
+        $prefs['vcsState'] = $state;
+        $user->setAttribute('prefs', $prefs);
+        $dbForConsole->updateDocument('users', $user->getId(), $user);
+        
+        $appName = App::getEnv('VCS_GITHUB_APP_NAME');
+        $response
+            ->addHeader('Cache-Control', 'no-store, no-cache, must-revalidate, max-age=0')
+            ->addHeader('Pragma', 'no-cache')
+            ->redirect("https://github.com/apps/$appName/installations/new");
+    });
+
+App::get('/v1/vcs/github/redirect')
+    ->desc('Capture installation and authorization from GitHub App')
+    ->groups(['api', 'vcs'])
+    ->label('scope', 'public')
+    ->label('error', __DIR__ . '/../../views/general/error.phtml')
+    ->param('installation_id', '', new Text(256), 'GitHub installation ID', true)
+    ->param('setup_action', '', new Text(256), 'GitHub setup actuon type', true)
+    // ->param('state', '', new Text(2048), 'GitHub state. Contains info sent when starting authorization flow.', true)
+    ->param('code', '', new Text(2048), 'OAuth2 code.', true)
+    ->inject('gitHub')
+    ->inject('user')
+    ->inject('project')
+    ->inject('request')
+    ->inject('response')
+    ->inject('dbForConsole')
+    ->action(function (string $installationId, string $setupAction, string $code, GitHub $github, Document $user, Document $project, Request $request, Response $response, Database $dbForConsole) {
+        // replace github url state with vcsState in user prefs attribute
```

Comment:
Because GitHub changed their API structure and don't have `state` param now. Since we need to map the installation with a project in Appwrite, we had to store it somewhere.

Reply:
Nessessary devil as of right now.
One day (few weeks ago) our integration stopped working. We noticed there is unexpected behaviour on Vercel and Netlify too.
We were aware Github screwed something. We didn't want to relay on them fixing it. We decided to use prefs as our own state management. We should probably keep it like this to not rely on github state. For future we might want better solution for such internal "prefs"

Reply:
For now we can probably prefix it with `$` too? 🤔 Anyway, the prefs are removed as soon as you get back to Appwrite after redirect.